File: src/ui/components/thinking-picker.ts
import {
  fuzzyFilter,
  getKeybindings,
  matchesKey,
  truncateToWidth,
  visibleWidth,
  type Component,
} from "@earendil-works/pi-tui";
import { theme } from "../../theme/theme.ts";

const LEVEL_DESCRIPTIONS: Record<string, string> = {
  off: "No reasoning",
  minimal: "Very brief reasoning (~1k tokens)",
  low: "Light reasoning (~2k tokens)",
  medium: "Moderate reasoning (~8k tokens)",
  high: "Deep reasoning (~16k tokens)",
  xhigh: "Extra-high reasoning (~32k tokens)",
  max: "Maximum reasoning",
};

interface ThinkingItem {
  level: string;
  description: string;
}

/**
 * Compact thinking-level picker styled like the model picker: a bordered list
 * with a `>` search row, a `✓` on the active level and no extra hint rows.
 */
export class ThinkingPicker implements Component {
  private filter = "";
  private selected = 0;
  private readonly items: ThinkingItem[];
  private filtered: ThinkingItem[];

  constructor(
    levels: string[],
    private current: string,
    private defaultLevel: string | undefined,
    private onSelect: (level: string) => void,
    private onSetDefault: (level: string) => void,
    private onCancel: () => void,
    private title = "Thinking",
  ) {
    this.items = levels.map((level) => {
      const full = LEVEL_DESCRIPTIONS[level] ?? "";
      // Show only the token-budget hint, e.g. "(~1k tokens)"; levels without one
      // (off/max) keep their short label.
      const hint = full.match(/\([^)]*\)/)?.[0] ?? full;
      return { level, description: hint };
    });
    this.filtered = this.items;
    const index = this.items.findIndex((item) => item.level === current);
    if (index >= 0) this.selected = index;
  }

  invalidate(): void {}

  private refilter(): void {
    this.filtered = this.filter.trim()
      ? fuzzyFilter(this.items, this.filter, (item) => `${item.level} ${item.description}`)
      : this.items;
    this.selected = Math.min(this.selected, Math.max(0, this.filtered.length - 1));
  }

  handleInput(data: string): void {
    const kb = getKeybindings();
    if (matchesKey(data, "escape") || matchesKey(data, "ctrl+c")) return this.onCancel();
    if (kb.matches(data, "app.thinking.save")) {
      const item = this.filtered[this.selected];
      if (item) this.onSetDefault(item.level);
      return;
    }
    if (matchesKey(data, "enter")) {
      const item = this.filtered[this.selected];
      if (item) this.onSelect(item.level);
      return;
    }
    if (matchesKey(data, "up")) {
      this.selected = this.filtered.length === 0 ? 0 : (this.selected + this.filtered.length - 1) % this.filtered.length;
      return;
    }
    if (matchesKey(data, "down")) {
      this.selected = this.filtered.length === 0 ? 0 : (this.selected + 1) % this.filtered.length;
      return;
    }
    if (matchesKey(data, "backspace")) {
      this.filter = this.filter.slice(0, -1);
      this.refilter();
      return;
    }
    if (data.length === 1 && data >= " " && data !== "\x7f") {
      this.filter += data;
      this.refilter();
    }
  }

  render(width: number): string[] {
    const t = theme();
    const total = Math.max(3, width);
    const inner = total - 2;
    const border = (text: string) => t.fg("borderAccent", text);
    const titleLabel = ` ${this.title} `;
    const title = total >= titleLabel.length + 6 ? titleLabel : "";
    const top =
      border("╭─") + t.fg("borderAccent", title) + border("─".repeat(Math.max(0, total - 3 - title.length)) + "╮");
    const rows: string[] = [];

    // ">" mirrors the selection arrow in the same column, like the model picker.
    const filterLine = "> " + t.fg("text", this.filter) + t.fg("accent", "█");
    rows.push(border("│") + " " + pad(truncateToWidth(filterLine, inner - 1, "…"), inner - 1) + border("│"));

    const visibleCount = 10;
    const count = Math.min(visibleCount, Math.max(1, this.filtered.length));
    const start =
      this.filtered.length <= visibleCount
        ? 0
        : Math.max(0, Math.min(this.selected - Math.floor(visibleCount / 2), this.filtered.length - visibleCount));
    for (let i = 0; i < count; i++) {
      const index = Math.max(0, start) + i;
      const item = this.filtered[index];
      if (!item) {
        rows.push(border("│") + " " + " ".repeat(inner - 1) + border("│"));
        continue;
      }
      const isSelected = index === this.selected;
      const isCurrent = item.level === this.current;
      const marker = isSelected ? t.fg("accent", "→ ") : "  ";
      // The active level is always green; the focused row's name is blue only
      // when it is not the active one.
      const name = isCurrent ? t.fg("success", item.level) : isSelected ? t.fg("accent", item.level) : t.fg("text", item.level);
      const suffix = item.level === this.defaultLevel ? `${item.description} · default` : item.description;
      const line = marker + name + t.fg("muted", `  ${suffix}`);
      rows.push(border("│") + " " + pad(truncateToWidth(line, inner - 1, "…"), inner - 1) + border("│"));
    }

    const bottom = border("╰" + "─".repeat(inner) + "╯");
    return [top, ...rows, bottom];
  }
}

function pad(text: string, width: number): string {
  return text + " ".repeat(Math.max(0, width - visibleWidth(text)));
}

File: internal/tui/thinking_picker.go
package tui

import (
	"strings"
	"unicode/utf8"

	tuitext "github.com/CarlvinceTan/midas/internal/tui/text"
	"github.com/CarlvinceTan/midas/pkg/ai"
)

type ThinkingPickerOptions struct {
	Model        ai.Model
	Current      ai.ThinkingLevel
	Default      ai.ThinkingLevel
	Title        string
	OnSelect     func(ai.ThinkingLevel)
	OnSetDefault func(ai.ThinkingLevel)
	OnCancel     func()
	Theme        *Theme
}

type thinkingPickerItem struct {
	level       ai.ThinkingLevel
	description string
}

type ThinkingPicker struct {
	FocusState
	items        []thinkingPickerItem
	filtered     []thinkingPickerItem
	filter       string
	current      ai.ThinkingLevel
	defaultLevel ai.ThinkingLevel
	title        string
	selected     int
	onSelect     func(ai.ThinkingLevel)
	onSetDefault func(ai.ThinkingLevel)
	onCancel     func()
	theme        *Theme
	rowTarget    map[int]int
}

var allThinkingLevels = []ai.ThinkingLevel{
	ai.ThinkingOff, ai.ThinkingMinimal, ai.ThinkingLow, ai.ThinkingMedium,
	ai.ThinkingHigh, ai.ThinkingXHigh, ai.ThinkingMax,
}

func ThinkingLevels(model ai.Model) []ai.ThinkingLevel {
	if !model.Reasoning {
		return []ai.ThinkingLevel{ai.ThinkingOff}
	}
	levels := make([]ai.ThinkingLevel, 0, len(allThinkingLevels))
	for _, level := range allThinkingLevels {
		mapped, exists := model.ThinkingLevelMap[level]
		if exists && mapped == nil {
			continue
		}
		levels = append(levels, level)
	}
	return levels
}

// NextThinkingLevel returns the next thinking level supported by model,
// wrapping to the first level after the last one.
func NextThinkingLevel(model ai.Model, current ai.ThinkingLevel) ai.ThinkingLevel {
	levels := ThinkingLevels(model)
	for index, level := range levels {
		if level == current {
			return levels[(index+1)%len(levels)]
		}
	}
	return levels[0]
}

func NewThinkingPicker(options ThinkingPickerOptions) *ThinkingPicker {
	descriptions := map[ai.ThinkingLevel]string{
		ai.ThinkingOff:     "No reasoning",
		ai.ThinkingMinimal: "(~1k tokens)",
		ai.ThinkingLow:     "(~2k tokens)",
		ai.ThinkingMedium:  "(~8k tokens)",
		ai.ThinkingHigh:    "(~16k tokens)",
		ai.ThinkingXHigh:   "(~32k tokens)",
		ai.ThinkingMax:     "Maximum reasoning",
	}
	levels := ThinkingLevels(options.Model)
	items := make([]thinkingPickerItem, 0, len(levels))
	selected := 0
	for index, level := range levels {
		items = append(items, thinkingPickerItem{level: level, description: descriptions[level]})
		if level == options.Current {
			selected = index
		}
	}
	title := strings.TrimSpace(options.Title)
	if title == "" {
		title = "Thinking"
	}
	defaultLevel := options.Default
	if defaultLevel == "" {
		defaultLevel = ai.ThinkingMedium
	}
	return &ThinkingPicker{
		items: items, filtered: append([]thinkingPickerItem(nil), items...), selected: selected,
		current: options.Current, defaultLevel: defaultLevel, title: title,
		onSelect: options.OnSelect, onSetDefault: options.OnSetDefault, onCancel: options.OnCancel,
		theme: options.Theme, rowTarget: make(map[int]int),
	}
}

// SetTitle is how Dialog names this picker's frame: the picker draws the frame, so
// the caller must not wrap it in a panel of its own.
func (p *ThinkingPicker) SetTitle(title string) {
	title = strings.TrimSpace(title)
	if title == "" {
		title = "Thinking"
	}
	p.title = title
}

func (p *ThinkingPicker) Invalidate()   {}
func (p *ThinkingPicker) Selected() int { return p.selected }

func (p *ThinkingPicker) refilter() {
	p.filtered = FuzzyFilter(p.items, p.filter, func(item thinkingPickerItem) string {
		return string(item.level) + " " + item.description
	})
	p.selected = min(p.selected, max(0, len(p.filtered)-1))
}

func (p *ThinkingPicker) HandleInput(data string) {
	if MatchesKey(data, "escape") || MatchesKey(data, "ctrl+c") {
		if p.onCancel != nil {
			p.onCancel()
		}
		return
	}
	if MatchesKey(data, "enter") {
		if p.selected < len(p.filtered) && p.onSelect != nil {
			p.onSelect(p.filtered[p.selected].level)
		}
		return
	}
	if MatchesKey(data, "up") {
		if len(p.filtered) > 0 {
			p.selected = (p.selected + len(p.filtered) - 1) % len(p.filtered)
		}
		return
	}
	if MatchesKey(data, "down") {
		if len(p.filtered) > 0 {
			p.selected = (p.selected + 1) % len(p.filtered)
		}
		return
	}
	if MatchesKey(data, "backspace") {
		if p.filter != "" {
			_, size := utf8.DecodeLastRuneInString(p.filter)
			p.filter = p.filter[:len(p.filter)-size]
			p.refilter()
		}
		return
	}
	if utf8.RuneCountInString(data) == 1 {
		r, _ := utf8.DecodeRuneInString(data)
		if r >= ' ' && r != 0x7f {
			p.filter += data
			p.refilter()
		}
	}
}

func (p *ThinkingPicker) HandleMouse(event MouseEvent) *MouseResult {
	if event.Type != MouseClick || event.Button != MouseLeft {
		return nil
	}
	index, ok := p.rowTarget[event.Y]
	if !ok || index >= len(p.filtered) {
		return nil
	}
	p.selected = index
	if p.onSelect != nil {
		p.onSelect(p.filtered[index].level)
	}
	render := true
	return &MouseResult{Handled: true, Render: &render}
}

func (p *ThinkingPicker) Render(width int) []string {
	total := max(3, width)
	inner := total - 2
	theme := p.theme
	if theme == nil {
		theme = CurrentTheme()
	}
	border := func(value string) string { return theme.FG("borderAccent", value) }
	titleLabel := " " + p.title + " "
	title := ""
	if total >= tuitext.VisibleWidth(titleLabel)+6 {
		title = titleLabel
	}
	top := border("╭─") + theme.FG("borderAccent", title) + border(strings.Repeat("─", max(0, total-3-tuitext.VisibleWidth(title)))+"╮")
	row := func(line string) string {
		contentWidth := max(1, inner-1)
		content := tuitext.TruncateToWidth(line, contentWidth, "…", true)
		content += strings.Repeat(" ", max(0, contentWidth-tuitext.VisibleWidth(content)))
		return border("│") + " " + content + border("│")
	}

	rows := []string{row("> " + theme.FG("text", p.filter) + theme.FG("accent", "█"))}
	const visible = 10
	count := min(visible, max(1, len(p.filtered)))
	start := 0
	if len(p.filtered) > visible {
		start = max(0, min(p.selected-visible/2, len(p.filtered)-visible))
	}
	p.rowTarget = make(map[int]int)
	for index := 0; index < count; index++ {
		position := start + index
		if position >= len(p.filtered) {
			rows = append(rows, row(""))
			continue
		}
		p.rowTarget[index+1] = position
		item := p.filtered[position]
		selected := position == p.selected
		marker := "  "
		if selected {
			marker = theme.FG("accent", "→ ")
		}
		name := string(item.level)
		if item.level == p.current {
			name = theme.FG("success", name)
		} else if selected {
			name = theme.FG("accent", name)
		} else {
			name = theme.FG("text", name)
		}
		suffix := item.description
		if item.level == p.defaultLevel {
			suffix += " · default"
		}
		rows = append(rows, row(marker+name+theme.FG("muted", "  "+suffix)))
	}
	bottom := border("╰" + strings.Repeat("─", inner) + "╯")
	return append(append([]string{top}, rows...), bottom)
}

func thinkingThemeName(level ai.ThinkingLevel) string {
	value := capitalizePanelTitle(strings.ToLower(string(level)))
	return "thinking" + value
}
